### PR TITLE
Include the project role of a request token or user in project API

### PR DESCRIPTION
**Pull request by ikhoon**
Motivation:

I think it is necessary to filter projects to which users belong or display their roles on the screen.

Currently, user information can't be obtained from `/api/v1/projects` API. So we need to call `/api/v1/projects/${projectName}` to get the metadata. That means 20 API calls should be made to list 20 projects in `/app/projects`. It will cause additional loads to servers and increase page loading time.

To resolve the N+1 query problem, I propose to add the role of a requested user to `ProjectDto`.

Modifications:

- Add `RepositoryListener` abstraction that can be continuously updated when a new change is pushed to a `Repository`. `Repository.watch()` has a similar function but it can be updated only once.
- Add `ProjectMetadata` to `DefaultProject` and use `RepositoryListener` to update `ProjectMetadata` in a project when new changes are made in `metadata.json`
- Expose user roles on the Projects page of the new web application.
- Refactor `ControlPlanePlugin` with `RepositoryListener` to simply the code.
- Miscellaneous) Show a loading indicator when a page is loading.

Result:

- The project role of a request token or user is now exposed in the project API.
- Users can see their roles on the projects page.
- Users can filter projects by their roles on the project page.

### CODE REVIEW

**Comment by ikhoon** [comment]
image
image

File: webapp/src/dogma/features/project/Projects.tsx [hunk 1]
```diff
@@ -107,37 +129,53 @@ export const Projects = () => {
...
+                      icon={<FcServices />}
+                      variant="ghost"
+                      colorScheme="teal"
+                      aria-label="Project Settings"
+                    />
+                  </Tooltip>
+                </ChakraLink>
+              );
+            } else {
+              // If the user is not a member of the project, do not show the project settings button.
```
Comment (minwoox) [inline webapp/src/dogma/features/project/Projects.tsx:162]: You meant owner?

File: webapp/src/dogma/features/project/Projects.tsx [hunk 2]
```diff
@@ -147,17 +185,22 @@ export const Projects = () => {
...
                     type="radio"
-                    onChange={(type) => dispatch(setProjectFilter(type as FilterType))}
+                    onChange={(type) => dispatch(setProjectFilter(type as ProjectFilterType))}
+                    alignItems={'center'}
                   >
-                    <MenuItemOption value="all">All</MenuItemOption>
-                    <MenuItemOption value="me">Created by me</MenuItemOption>
+                    <MenuItemOption value="ALL">All</MenuItemOption>
+                    <MenuItemOption value="MEMBER">I am a member</MenuItemOption>
+                    <MenuItemOption value="CREATOR">Created by me</MenuItemOption>
```
Comment (minwoox) [inline webapp/src/dogma/features/project/Projects.tsx:203]: Do you have any easy way to test the UI in local?

File: xds/src/main/java/com/linecorp/centraldogma/xds/internal/ControlPlanePlugin.java [hunk 3]
```diff
@@ -174,52 +174,25 @@ private void setXdsResources(String path, String contentAsText, String repoName)
...
-                                   watchingXdsProjects.remove(repoName);
-                                   return;
-                               }
-                               watchXdsProject(repository, Revision.INIT);
-                           });
-                           // Watch dogma repository again to catch up newly created xDS projects.
-                           watchDogmaRepository(repositoryManager, entry.revision());
-                           return null;
-                       }, CONTROL_PLANE_EXECUTOR);
+        dogmaRepository.addListener(RepositoryListener.of(MetadataService.METADATA_JSON, entries -> {
```
Comment (minwoox) [inline xds/src/main/java/com/linecorp/centraldogma/xds/internal/ControlPlanePlugin.java:180]: 👍

File: xds/src/main/java/com/linecorp/centraldogma/xds/internal/ControlPlanePlugin.java [hunk 4]
```diff
@@ -331,7 +304,8 @@ public void onV3StreamDeltaRequest(long streamId, DeltaDiscoveryRequest request)
                 throws RequestException {}
 
         @Override
-        public void onV3StreamResponse(long streamId, DiscoveryRequest request, DiscoveryResponse response) {
+        public void onV3StreamResponse(long streamId, DiscoveryRequest request,
+                                       DiscoveryResponse response) {
```
Comment (minwoox) [inline xds/src/main/java/com/linecorp/centraldogma/xds/internal/ControlPlanePlugin.java:308]: revert?

**Review by minwoox — COMMENTED** [review]
Looks promising! 👍

File: server/src/main/java/com/linecorp/centraldogma/server/internal/api/ProjectServiceV1.java [hunk 5]
```diff
@@ -83,9 +88,58 @@ public CompletableFuture<List<ProjectDto>> listProjects(@Param @Nullable String
...
+
+    private ProjectRoleDto getUserRole(Project project, User user) {
+        if (user.isAdmin()) {
+            return ProjectRoleDto.OWNER;
+        }
+
+        final ProjectMetadata metadata = project.metadata();
+        if (metadata == null) {
+            // Metadata is null for the internal project which belongs to administrators.
+            return user.isAdmin() ? ProjectRoleDto.OWNER : ProjectRoleDto.GUEST;
```
[the remark this answers is not in the document]
Comment (ikhoon) [reply server/src/main/java/com/linecorp/centraldogma/server/internal/api/ProjectServiceV1.java:106]: Can't admin own `dogma` or internal repos?

File: webapp/src/dogma/features/project/Projects.tsx (hunk 2 above)
Comment (ikhoon) [reply webapp/src/dogma/features/project/Projects.tsx:203]: I prefer the following steps:

- Run `ShiroCentralDogmaTestServer.main()` using IntelliJ which is faster than `./gradlew runTestShiroServer`.
- If the server starts successfully, execute `npm run develop` and access `http://127.0.0.1:3000`.

If `test` is meant to run a unit test, we can run `npm run test`. I temporarily disable tests that fetch data from a mock server after upgrading dependencies. Because there were compatibility issues that I haven't solved yet.

File: server/src/main/java/com/linecorp/centraldogma/server/internal/api/ProjectServiceV1.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (ikhoon) [reply server/src/main/java/com/linecorp/centraldogma/server/internal/api/ProjectServiceV1.java:106]: Nevermind. Now I understood.

File: webapp/src/dogma/features/project/Projects.tsx (hunk 2 above)
Comment (minwoox) [reply webapp/src/dogma/features/project/Projects.tsx:203]: > Run ShiroCentralDogmaTestServer.main() using IntelliJ which is faster than ./gradlew runTestShiroServer.
If the server starts successfully, execute npm run develop and access http://127.0.0.1:3000.

That's what I was looking for. 😉

File: common/src/main/java/com/linecorp/centraldogma/internal/api/v1/ProjectDto.java [hunk 6]
```diff
@@ -24,36 +24,58 @@
...
     }
 
-    public ProjectDto(String name, Author creator, long creationTimeMillis) {
+    @VisibleForTesting
+    @JsonCreator
+    public ProjectDto(@JsonProperty("name") String name,
+                      @JsonProperty("creator") @Nullable Author creator,
+                      @JsonProperty("url") @Nullable String url,
+                      @JsonProperty("userRole") @Nullable ProjectRole userRole,
+                      @JsonProperty("createdAt") @Nullable String createdAt) {
```
Comment (jrhee17) [inline common/src/main/java/com/linecorp/centraldogma/internal/api/v1/ProjectDto.java:65]: Question) Even in tests, is there a need for the nullable contract since server-side always seems to fill in these fields?

File: common/src/main/java/com/linecorp/centraldogma/internal/api/v1/ProjectDto.java [hunk 7]
```diff
@@ -74,9 +96,9 @@ public String url() {
     }
 
     @Nullable
-    @JsonProperty("reposUrl")
```
Comment (jrhee17) [inline common/src/main/java/com/linecorp/centraldogma/internal/api/v1/ProjectDto.java:77]: note: checked that this field doesn't seem to be used in the previous UI also, which guarantees backwards compatibility

File: server/src/main/java/com/linecorp/centraldogma/server/internal/api/TokenService.java [hunk 8]
```diff
@@ -103,7 +104,7 @@ public CompletableFuture<Collection<Token>> listTokens(User loginUser) {
     @StatusCode(201)
     @ResponseConverter(CreateApiResponseConverter.class)
     public CompletableFuture<HttpResult<Token>> createToken(@Param String appId,
-                                                            @Param boolean isAdmin,
+                                                            @Param @Default("false") boolean isAdmin,
```
Comment (jrhee17) [inline server/src/main/java/com/linecorp/centraldogma/server/internal/api/TokenService.java:107]: Question) Is this related to changes in this PR? Or is this a separate improvement?

File: server/src/main/java/com/linecorp/centraldogma/server/internal/storage/project/DefaultProject.java [hunk 9]
```diff
@@ -196,11 +199,63 @@ private void initializeMetadata(long creationTimeMillis, Author author) {
...
+            if (lastRevision.compareTo(lastMetadataRevision) <= 0) {
+                // An old data.
+                return;
+            }
+
+            try {
+                final ProjectMetadata projectMetadata = Jackson.treeToValue(entry.content(),
+                                                                            ProjectMetadata.class);
+                lastMetadataRevision = lastRevision;
+                DefaultProject.this.projectMetadata = projectMetadata;
```
Comment (jrhee17) [inline server/src/main/java/com/linecorp/centraldogma/server/internal/storage/project/DefaultProject.java:254]:
```suggestion
                this.projectMetadata = projectMetadata;
```

File: server/src/main/java/com/linecorp/centraldogma/server/internal/api/ProjectServiceV1.java [hunk 10]
```diff
@@ -83,9 +87,47 @@ public CompletableFuture<List<ProjectDto>> listProjects(@Param @Nullable String
...
+                                    .collect(toImmutableList());
+        }, executor);
+    }
+
+    private static ProjectRole getUserRole(Project project, User user) {
+        if (user.isAdmin()) {
+            return ProjectRole.OWNER;
+        }
+
+        final ProjectMetadata metadata = project.metadata();
```
Comment (jrhee17) [inline server/src/main/java/com/linecorp/centraldogma/server/internal/api/ProjectServiceV1.java:102]: Question) I believe `Project#metadata` is updated asynchronously. This means if a user changes his/her metadata and directly navigates to `/projects`, I think it is possible that the change may not be reflected depending on whether the `find` in the listener has been invoked or not.

I think it should be fine since listing "my projects" can be considered non-critical, but just wanted to make sure if this is an intentional design choice as opposed to just using a write-through cache. (I haven't checked if there are any UI paths which update metadata and navigate to `/projects` immediately)

**Review by jrhee17 — APPROVED** [review]
Left some minor comments but looks good to me 👍 👍 👍

File: common/src/main/java/com/linecorp/centraldogma/common/ProjectRole.java [hunk 11]
```diff
@@ -14,7 +14,7 @@
  * under the License.
  */
 
-package com.linecorp.centraldogma.server.metadata;
+package com.linecorp.centraldogma.common;
```
[the remark this answers is not in the document]
Comment (ikhoon) [reply common/src/main/java/com/linecorp/centraldogma/common/ProjectRole.java:17]: Added `breaking change` label.

File: server/src/main/java/com/linecorp/centraldogma/server/storage/repository/RepositoryListener.java [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+package com.linecorp.centraldogma.server.storage.repository;
+
+import static java.util.Objects.requireNonNull;
+
+import java.util.Map;
+
+import com.linecorp.centraldogma.common.Entry;
+import com.linecorp.centraldogma.common.Query;
+
+import io.netty.util.internal.shaded.org.jctools.queues.MessagePassingQueue.Consumer;
```
[the remark this answers is not in the document]
Comment (ikhoon) [reply server/src/main/java/com/linecorp/centraldogma/server/storage/repository/RepositoryListener.java:26]: 😱

File: server/src/main/java/com/linecorp/centraldogma/server/internal/api/ProjectServiceV1.java (hunk 10 above)
Comment (ikhoon) [reply server/src/main/java/com/linecorp/centraldogma/server/internal/api/ProjectServiceV1.java:102]: Good point. As you know, the consistent model of Central Dogma is eventual consistency and does not provide a transaction mechanism. 

Even before this change, users may see old content right after a commit is pushed successfully. For example, `ReplicationLagTolerantCentralDogma` is a workaround to make up for the downside of eventual consistency.

File: server/src/main/java/com/linecorp/centraldogma/server/internal/api/TokenService.java (hunk 8 above)
Comment (ikhoon) [reply server/src/main/java/com/linecorp/centraldogma/server/internal/api/TokenService.java:107]: It is not related to this PR.
I've fixed it in this PR together since it looked like a trivial change.

File: common/src/main/java/com/linecorp/centraldogma/internal/api/v1/ProjectDto.java (hunk 6 above)
Comment (ikhoon) [reply common/src/main/java/com/linecorp/centraldogma/internal/api/v1/ProjectDto.java:65]: This constructor was added for deserializing the API responses in tests.
https://github.com/line/centraldogma/blob/11e03be535ec2f0c3bd942f161169f8b7af24eb4/server/src/test/java/com/linecorp/centraldogma/server/internal/api/ProjectServiceV1ListProjectTest.java#L339-L340

It wasn't used for `?status=removed` responses but made them nullable for future usage.

Comment (jrhee17) [reply common/src/main/java/com/linecorp/centraldogma/internal/api/v1/ProjectDto.java:65]: > This constructor was added for deserializing the API responses in tests.

I understand that. I meant even in the test you linked aren't the fields always being filled in (non-null) anyways?

Comment (ikhoon) [reply common/src/main/java/com/linecorp/centraldogma/internal/api/v1/ProjectDto.java:65]: > you linked aren't the fields always being filled in (non-null) anyways?

Right. They are non-null for normal projects. The values are nullable for removed projects.
https://github.com/line/centraldogma/blob/c0673c169c9a19034719c5a0e2594bc85415a577/server/src/main/java/com/linecorp/centraldogma/server/internal/api/ProjectServiceV1.java#L84-L86

File: common/src/main/java/com/linecorp/centraldogma/internal/api/v1/ProjectDto.java [hunk 13]
```diff
@@ -24,36 +24,58 @@
...
+        this.userRole = userRole;
+        this.createdAt = createdAt;
+    }
+
+    public ProjectDto(String name, Author creator, ProjectRole userRole, long creationTimeMillis) {
         this.name = requireNonNull(name, "name");
         this.creator = requireNonNull(creator, "creator");
         createdAt = ISO_INSTANT.format(Instant.ofEpochMilli(creationTimeMillis));
         url = PROJECTS_PREFIX + '/' + name;
+        this.userRole = userRole;
```
Comment (minwoox) [inline common/src/main/java/com/linecorp/centraldogma/internal/api/v1/ProjectDto.java:78]: rnn?

File: server/src/main/java/com/linecorp/centraldogma/server/internal/storage/project/DefaultProject.java [hunk 14]
```diff
@@ -196,11 +199,63 @@ private void initializeMetadata(long creationTimeMillis, Author author) {
...
                                                                  ImmutableMap.of(),
                                                                  userAndTimestamp, null);
+            final CommitResult result =
+                    dogmaRepo.commit(headRev, creationTimeMillis, Author.SYSTEM,
+                                     "Initialize metadata", "",
+                                     Markup.PLAINTEXT,
+                                     Change.ofJsonUpsert(METADATA_JSON, Jackson.valueToTree(metadata)))
+                             .join();
+            lastMetadataRevision = result.revision();
+            this.projectMetadata = metadata;
```
Comment (minwoox) [inline server/src/main/java/com/linecorp/centraldogma/server/internal/storage/project/DefaultProject.java:209]:
```suggestion
            projectMetadata = metadata;
```

File: server/src/main/java/com/linecorp/centraldogma/server/internal/storage/project/DefaultProject.java [hunk 15]
```diff
@@ -89,15 +98,19 @@ public class DefaultProject implements Project {
...
                 creationTimeMillis = creation.timestampMillis();
                 author = Author.ofEmail(creation.user());
+                attachMetadataListener();
             } else {
                 creationTimeMillis = repos.get(REPO_DOGMA).creationTimeMillis();
                 author = repos.get(REPO_DOGMA).author();
             }
             success = true;
+        } catch (Exception e) {
+            throw new CentralDogmaException("failed to initialize internal repositories", e);
```
Comment (minwoox) [inline server/src/main/java/com/linecorp/centraldogma/server/internal/storage/project/DefaultProject.java:113]: nit:
```suggestion
            throw new CentralDogmaException("failed to initialize internal repositories of " + name, e);
```
Please also add the project name to the exception message below.

File: server/src/main/java/com/linecorp/centraldogma/server/internal/storage/project/DefaultProject.java [hunk 16]
```diff
@@ -196,11 +199,63 @@ private void initializeMetadata(long creationTimeMillis, Author author) {
...
+
+        final Repository dogmaRepo = repos.get(REPO_DOGMA);
+        dogmaRepo.addListener(RepositoryListener.of(Query.ofJson(METADATA_JSON), entry -> {
+            if (entry == null) {
+                logger.warn("{} file is missing in {}/{}", METADATA_JSON, name, REPO_DOGMA);
+                return;
+            }
+
+            final Revision lastRevision = entry.revision();
+            if (lastRevision.compareTo(lastMetadataRevision) <= 0) {
```
Comment (minwoox) [inline server/src/main/java/com/linecorp/centraldogma/server/internal/storage/project/DefaultProject.java:245]: nit: `assert lastMetadataRevision != null;`
We might want to assign it to a local variable not to access the field twice.

**Review by minwoox — APPROVED** [review]
Thanks! 👍 👍 👍

File: server/src/main/java/com/linecorp/centraldogma/server/internal/api/ProjectServiceV1.java (hunk 10 above)
Comment (jrhee17) [reply server/src/main/java/com/linecorp/centraldogma/server/internal/api/ProjectServiceV1.java:102]: I see, I understood that even if a user creates a project and directly navigates to the main page he/she may not see her project highlighted.

Comment (ikhoon) [reply server/src/main/java/com/linecorp/centraldogma/server/internal/api/ProjectServiceV1.java:102]: When I tested, I was able to see newly created projects in most cases. But it may happen if there is a significant delay. Let's leave it a known issue for now.
